### ISSUE TITLE
Half migrated users can not be deleted in Keycloak

### ISSUE THREAD

**Issue by DaBonse** [daniel-frak/keycloak-user-migration#185]
Keycloak 25.0.1 | 25.0.2

Plugin Version 5.0.0

Hi,

I have found a strange situation I can not explain.

When a not migrated user starts the password reset flow and not finished the flow, then the user is migrated without password. This is OK and we could handle it with a clean up after script.

But this user can not be deleted in Keycloak. Keycloak Admin UI failed with an error. Deleting over Admin API also failes with status code 400, and error message: User couldn't be deleted. In the Keycloak logs, I didn't found a reason why the user can not be deleted. A look at the database also gives me no reason for this behaviour. I have expected the user is read only or something else.

Strange is, a user created in Keycloak without password can be deleted. 
A half plugin migrated user can be deleted, after setting a password by admin.

This is a problem for our clean up script, because we can not delete the user without setting a password. What makes no sense.

Is this a Bug or is there a good reason why the user can not be deleted?

Thanks for helping,

DaBonse

**Comment by DaBonse** [comment]
Have anyone an explanation for this?

**Comment by sergevic** [comment]
I have experienced the same problem. It can be reproduced with these steps:
1. try to login with username and wrong password to the keycloak. username exists in the legacy system.
2. user federation plugin creates the user, but with no credentials. as it is described in the readme ("Example REST client behavior" chapter)
3. a call to Admin API to remove the user fails with "User couldn't be deleted" message. Deletion via Keycloak Admin UI also fails.
4. add a password to the user credentials in the keycloak
5. call to Admin API to remove the user succeeds, as well as deletion via Keycloak Admin UI

Why does the plugin create a user even if the password is incorrect? The logic a bit strange for me.

**Comment by daniel-frak** [comment]
I looked into this, and it seems that it's a Keycloak bug. I have opened an issue in the Keycloak repository:

https://github.com/keycloak/keycloak/issues/34747

To briefly explain what's going on, the migration plugin creates a federation link when a user first tries to log in. This federation link is then severed once the user logs in successfully for the first time. Therefore, if the user tries to log in for the first time with an incorrect password, they will be added to Keycloak with the federation link present.

The bug, then, appears to be Keycloak not allowing users with active federation links to be deleted.

A workaround is to go to the plugin page in the admin UI and use the "Remove imported" action, which will remove all users with unsevered federation links.

@sergevic I believe the early creation of the user (with the federation link) is necessary for things like the "forgot password" flow to work correctly.

**Comment by daniel-frak** [comment]
Hello again,

It seems this issue was caused by a quirk of design in Keycloak, where user deletion in this case won't work correctly unless `UserRegistrationProvider` is implemented on the provider. I don't necessarily agree with Keycloak's developers that this is the best design, but I've implemented NO-OP methods from `UserRegistrationProvider` in `LegacyProvider` and it seems half-migrated user deletion works fine now.

I'm closing this as fixed. Let me know if you have any issues with it starting from v6.0.1 of the plugin.
